flang/lib/Lower/IO.cpp

```diff
@@ -287,10 +263,9 @@ genOutputItemList(Fortran::lower::AbstractConverter &converter,
     auto argType = outputFunc.getType().getInput(1);
     llvm::SmallVector<mlir::Value, 3> outputFuncArgs = {cookie};
     Fortran::lower::CharacterExprHelper helper{builder, loc};
-    if (!isFormatted) {
+    if (argType.isa<fir::BoxType>()) {
```

Comment:
Is this always going to be true when the I/O statement is unformatted? It's a bug if it is ever possible to have an unformatted I/O case and to not forcibly convert the argument to be boxed.

Reply:
The responsibility of deciding if the argument must be boxed is left to `getOutputFunc` above that is taking the `isFormatted` argument.
With the current code, we could have ended-up doing weird things like making a descriptor and casting it to an `i8*` if `getOutputFunc` and this code ever started to disagree in their logic (bug). It seems safer to me to have a single decision point (`getOutputFunc`) and let the code here follow the decision rather than making the, hopefully same, decision again.

Reply:
getOutputFunc is just returning the OutputDescriptor signature in that case. So that should be a box, yes.

What I'm wondering about here though is maybe an `assert` sort of check to double-check at this point that we're still going down the expected path. If someone changed the header file in runtime for some reason, this could go down an entirely unexpected path, no?

Reply:
Sure, I added an assert to check that.